src/envs/tools.py

```diff
@@ -0,0 +1,643 @@
+            num_tasks=CFG.num_train_tasks,
+            num_items_lst=CFG.tools_num_items_train,
+            num_contraptions_lst=CFG.tools_num_contraptions_train,
+            rng=self._train_rng)
+
+    def get_test_tasks(self) -> List[Task]:
+        return self._get_tasks(
+            num_tasks=CFG.num_test_tasks,
+            num_items_lst=CFG.tools_num_items_test,
+            num_contraptions_lst=CFG.tools_num_contraptions_test,
+            rng=self._test_rng)
+
+    @property
+    def predicates(self) -> Set[Predicate]:
+        return {
+            self._HandEmpty, self._HoldingScrew, self._HoldingScrewdriver,
+            self._HoldingNail, self._HoldingHammer, self._HoldingBolt,
+            self._HoldingWrench, self._ScrewOnTable, self._NailOnTable,
+            self._BoltOnTable, self._ScrewPlaced, self._NailPlaced,
+            self._BoltPlaced, self._ScrewFastened, self._NailFastened,
+            self._BoltFastened, self._ScrewdriverGraspable,
+            self._HammerGraspable
+        }
+
+    @property
+    def goal_predicates(self) -> Set[Predicate]:
+        return {
+            self._ScrewPlaced, self._NailPlaced, self._BoltPlaced,
+            self._ScrewFastened, self._NailFastened, self._BoltFastened
+        }
+
+    @property
+    def types(self) -> Set[Type]:
+        return {
+            self._robot_type, self._screw_type, self._screwdriver_type,
+            self._nail_type, self._hammer_type, self._bolt_type,
+            self._wrench_type, self._contraption_type
+        }
+
+    @property
+    def options(self) -> Set[ParameterizedOption]:
+        return {
+            self._PickScrew, self._PickScrewdriver, self._PickNail,
+            self._PickHammer, self._PickBolt, self._PickWrench, self._Place,
+            self._FastenScrewWithScrewdriver, self._FastenNailWithHammer,
+            self._FastenBoltWithWrench, self._FastenScrewByHand
+        }
+
+    @property
+    def action_space(self) -> Box:
+        # Actions are 4-dimensional vectors: [x, y, is_pick bit, is_place bit]
+        return Box(
+            np.array([self.table_lx, self.table_ly, 0, 0], dtype=np.float32),
+            np.array([self.table_ux, self.table_uy, 1, 1], dtype=np.float32))
+
+    def render(self,
+               state: State,
+               task: Task,
+               action: Optional[Action] = None) -> List[Image]:
+        raise NotImplementedError
+
+    def _get_tasks(self, num_tasks: int, num_items_lst: List[int],
+                   num_contraptions_lst: List[int],
+                   rng: np.random.Generator) -> List[Task]:
+        tasks = []
+        for i in range(num_tasks):
+            num_items = num_items_lst[i % len(num_items_lst)]
+            num_contraptions = num_contraptions_lst[i %
+                                                    len(num_contraptions_lst)]
+            data = {}
+            # Initialize robot
+            data[self._robot] = np.array([1.0])  # fingers start off open
```

Comment:
add dtype=np.float32? idk if necessary but it's used everywhere else

Reply:
actually i copied this from blocks/painting where it's not used there lol. but i just fixed those too